### ISSUE TITLE
Ionic 4 - iOS accessory bar next/previous buttons disabled using ion-item

### ISSUE THREAD

**Issue by KevinKelchen** [ionic-team/ionic-framework#16498]
# Bug Report

**Ionic Info**
```
Ionic:

   ionic (Ionic CLI)             : 4.6.0 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.0.0-rc.2
   @angular-devkit/build-angular : 0.12.2
   @angular-devkit/schematics    : 7.2.2
   @angular/cli                  : 7.2.2
   @ionic/angular-toolkit        : 1.2.2

Cordova:

   cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms     : ios 4.5.5
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 2.3.2, (and 4 other plugins)

System:

   Android SDK Tools : 26.1.1 (/Users/kelchen/Library/Android/sdk)
   ios-deploy        : 2.0.0
   ios-sim           : 7.0.0
   NodeJS            : v10.13.0 (/usr/local/bin/node)
   npm               : 6.4.1
   OS                : macOS Mojave
   Xcode             : Xcode 10.1 Build version 10B61
```

**Describe the Bug**
When using the `cordova-plugin-ionic-keyboard` Cordova plugin on iOS with the keyboard accessory bar enabled, the next and previous buttons of the accessory bar are disabled for navigating between form inputs that are not child to the same `ion-item` element.

**Steps to Reproduce**
1. Start a new project with `ionic start keyboard blank --type=angular`.
    - _Note: You may need to edit the `package.json` to upgrade it from Ionic 4 RC1 to RC2 if the starter templates have not yet been updated for RC2._
2. Path into the project directory and run `ionic cordova prepare ios`. Say `Y` to the prompt to install the iOS platform. `cordova-plugin-ionic-webview` should then be added to the project.
3. Run `ionic cordova prepare ios` again. It will download/install the Cordova plugins that are part of the template.
4. In `app.component.ts` after the platform is ready and when the `StatusBar` and `SplashScreen` plugins are being used, add the following line to enable the keyboard accessory bar:
```typescript
(window as any).Keyboard.hideFormAccessoryBar(false);
```
5. In `home.page.html` inside of `ion-content` add some adjacent `ion-input` elements as well as some adjacent native HTML `input` elements like this:
```html
<ion-content padding>
  The world is your oyster.
  <p>If you get lost, the <a target="_blank" rel="noopener" href="https://beta.ionicframework.com/docs/">docs</a> will be your guide.</p>
  <input />
  <input />
  <ion-item>
    <input />
    <ion-input></ion-input>
  </ion-item>
  <ion-input></ion-input>
  <ion-input></ion-input>
</ion-content>
```
6. Run `ionic cordova run ios --device`. This will build the app and deploy it to a device.
7. When the app is up and running on the device, tap on the first native HTML `input` element to put focus in the input. The keyboard will open. Notice on the keyboard accessory bar that the next button can be used to navigate to the second native `input` element. Tap the next button on the keyboard accessory bar to move to the second `input` element. You'll see the previous button also enable so you can move to the previous input.
8. With focus on the second native `input` element, tap the next button on the keyboard accessory bar. Notice that you jump over both the native `input` as well as the `ion-input` elements that are child to the `ion-item` element. The element that is in focus is actually the first of the two `ion-input`s that are direct children to `ion-content`. You can then also use the keyboard accessory bar's next button to navigate to the second `ion-input` that's a direct child to `ion-content`.

**Related Code**
The steps to reproduce above should be all that you need (using the starter template and following the steps).

**Expected Behavior**
I expected to be able to use the iOS keyboard accessory bar previous/next buttons to navigate between form inputs that are not child to the same `ion-item` element, which I could do in Ionic 3.

**Additional Context**
Originally this issue focused on the Shadow DOM of `ion-input` contributing to the keyboard next/previous buttons not working; `ion-item` wasn't in the initial sample to reproduce or noted as being a contributing factor. With Shadow DOM being removed on `ion-input` in RC2 I tried once again.

While the initial sample to reproduce then worked, our application did not. I determined this is because in our application we wrap our `ion-input` elements with `ion-item` elements. I've updated this issue accordingly.

**Comment by paulstelzer** [comment]
Ionic uses shadow dom so I think you have to write this logic yourself. Or do you have an idea to fix it?

**Comment by KevinKelchen** [comment]
Thanks for responding, @paulstelzer! 🙂

At this point I do not know how to resolve the issue. I am not sure if the problem is due to ShadowDOM, another reason in Ionic 4 (e.g., `ion-input` was somewhat re-written in v4 (Angular ➡️ Stencil) so there could be greater potential for bugs given the churn), or something in the [`cordova-plugin-ionic-keyboard` Cordova plugin](https://github.com/ionic-team/cordova-plugin-ionic-keyboard); ShadowDOM is just my suspicion. I have not dug in to confirm what the underlying issue is. Have you done so? At this point I've only submitted the issue figuring those who better understand Ionic would be more efficient and effective assessing the cause and potential solutions.

If I were to start looking into it myself I would probably try to understand how the keyboard determines it needs to enable/disable the next/previous buttons. Not sure if that would be in the plugin code or is something internal to Apple's keyboard. I will try taking a look and can see where I get, but I would appreciate help from anyone more knowledgeable.

I would expect this functionality to be supported out of the box as it involves Ionic's own keyboard plugin supporting Ionic's own, most basic, Ionic form element, rather than logic each user's application would be required to implement if they want two-thirds of the accessory bar functions to work and for the bar to do anything besides provide a `Done` button. 🙂 However, if it's determined that the logic belongs in the application code instead of the framework/plugin, does someone have an example or ideas they are willing to share?

Thank you!

**Comment by paulstelzer** [comment]
Mh, good question, but I cannot answer it right now. I label it as help wanted, maybe someone with more knowledge than we both has an idea if 1) this is possible in `@ionic/core` or the Ionic cordova keyboard plugin :)

**Comment by paulstelzer** [comment]
I close this for now because it should fixed with #17043 . After rc.2 release you can check again and if it still not working, just create a comment here and I reopen

**Comment by paulstelzer** [comment]
I think this is a browser issue. In Chrome and Firefox everything is working right, only Safari has problems for me

**Comment by paulstelzer** [comment]
The issue affects normal browser (in PWA) and native (using cordova plugin keyboard). If it could fixed in normal browser, it should fixed in native, too (hopefully)

**Comment by abennouna** [comment]
I have the same issue in a PWA in **iOS 12.1.2**, maybe this can help the diagnosis:
- in **Safari** & **Firefox**, the previous/next buttons are disabled
![safari-ios-disabled-next-prev](https://user-images.githubusercontent.com/1470318/51550340-16d2fe00-1e64-11e9-928b-dc259d9baa28.gif)

- in **Chrome**, they are enabled, but they do nothing
![chrome-ios-nonworking-next-previous](https://user-images.githubusercontent.com/1470318/51550334-13d80d80-1e64-11e9-8e87-ab738c7cca05.gif)

**Comment by wallstudios** [comment]
Any progress with this? Using a normal browser (firefox) I can't tab between inputs if they are wrapped in ion-items. 

eg
`                <ion-col>
                  <ion-item>
                    <ion-label position="stacked">First Name</ion-label>
                    <ion-input tabindex="2" formControlName="first_name" ></ion-input>
                  </ion-item>
                </ion-col>
                <ion-col>
                  <ion-item>
                    <ion-label position="stacked">Last Name</ion-label>
                    <ion-input tabindex="3" formControlName="last_name" ></ion-input>
                  </ion-item>
                </ion-col>`

**Comment by thmclellan** [comment]
I created a simple test repo and confirmed on both Firefox and iOS 11 Safari: 

- `<ion-item>` is breaking iOS accessory next/back bar and FireFox tabbing behaviour for both `<ion-input>` and plain HTML `<input>` elements 
- `<ion-input>`s without `<ion-item>` are working fine 

It seems like iOS Safari and Firefox are handling inputs inside shadow DOM elements differently from an accessibility standpoint. 

@manucorporat did a lot of work fixing keyboard issues with ion-input a month ago... I'm not sure the implications of disabling shadow dom from `ion-item` or if there's another work around. 

Anyway here's the test repo in case it helps: https://github.com/thmclellan/ionic4-tab-test

![image](https://user-images.githubusercontent.com/9682900/52352675-e487db80-29e1-11e9-9c12-c8ddb7a77d36.png)

This `<ion-item>` issue seems to be the root cause of related Firefox tabs issue https://github.com/ionic-team/ionic/issues/16927.

**Comment by wallstudios** [comment]
Just downloaded Firefox nightly 67 and the tab order in working. I suspect it is a Firefox issue.

![image](https://user-images.githubusercontent.com/8520022/52380252-fd18f400-2abc-11e9-8bdc-41b5ec8ed020.png)

**Comment by AndrinGautschi** [comment]
Is there something new about this issue?

**Comment by touch2hussain** [comment]
Any update on this issue?, Pages like a sign up with too many input fields sucks in ios/safari :(

**Comment by kensodemann** [comment]
I am working with someone who is likely going to need this. I created a simple repo that demos the issue: https://github.com/kensodemann/test-keyboard

**Comment by nporchis** [comment]
I also have the same issue in pwa with safari on iOS (V12.1.1 and 12.3.1). On android with chrome, the next button of the keyword is working well. Any update would be much appreciated :)

**Comment by NiklasMerz** [comment]
This is still an issue with `@ionic/angular` 4.11.3 on Cordova and iOS 13.3

**Comment by cyptus** [comment]
would be awesome to see some improvments here

**Comment by tomaszDziedzic2** [comment]
Also need this feature, I would be grateful if you could fix it.

**Comment by manish-sh14** [comment]
even i am looking into this, did not get the solution yet. next/previous button won't work on ion-input if its wrap in ion-item.

**Comment by cyptus** [comment]
@liamdebeasi may you got infos about this liam?

**Comment by justjoeyuk** [comment]
This is quite a big issue...

**Comment by liamdebeasi** [comment]
Thanks for the issue. Wanted to provide a brief update here.

This is a known issue in Webkit/Safari. We have reported the issue to the WebKit team (https://bugs.webkit.org/show_bug.cgi?id=203292), and they are aware of the issue and are investigating. The issue stems from having inputs in the shadow DOM. In the case of Ionic, putting `ion-input` elements in each of their own separate `ion-item` elements causes this issue to occur.

I will provide an update here when I have more to share. Thanks!

**Comment by visant** [comment]
Can you not just change the shadow property to the scope property in the ion-item, as you have done it for #17020 (PR #17043)? I believe the issue here very is similar.

**Comment by liamdebeasi** [comment]
Removing `ion-item` from the Shadow DOM would result in features being lost from the component (such as slots), and I do not think that is a good trade off.

**Comment by ipehimanshu** [comment]
@liamdebeasi 
is there any update for this issue ?

**Comment by liamdebeasi** [comment]
Hi everyone,

We are waiting on the WebKit team to resolve this issue. I am going to lock this thread for now since we have all the required information we need, but I will post an update here when this is resolved from the WebKit side. Thanks!

**Comment by liamdebeasi** [comment]
Hi everyone,

I can confirm that this issue is resolved in iOS 14.5 beta 1. I will follow up here when iOS 14.5 ships for all iOS users.

**Comment by liamdebeasi** [comment]
Hi there,

iOS 14.5 just shipped, and it includes a fix for this issue.

Thanks for everyone's patience!
